### PR TITLE
Add session RPC wrappers and close test coverage gaps

### PR DESCRIPTION
**Pull request by krukow**
## Summary

Close test coverage gaps identified by a file-by-file parity review against the Node.js and Python SDKs, and add named wrapper functions for all remaining generated session RPC methods.

### Test coverage gaps closed (27 new tests)

| Area | Tests | What was missing |
|---|---|---|
| **Hooks** | 6 | All 6 hook types (preToolUse, postToolUse, sessionStart, etc.) + edge cases |
| **User input** | 2 | `userInput.request` server→client RPC handler |
| **System message transforms** | 3 | Transform fn callbacks, error fallback, passthrough |
| **Tool result normalization** | 3 | String, nil, and structured `ToolResultObject` via v3 broadcast |
| **Session RPC wrappers** | 12 | Integration tests for all new functions |
| **MCP config** | 1 | `mcp.config.list` server-level RPC |

### New experimental RPC wrappers (19 functions)

**Session-level** (`session.clj`):
- `mode-get`, `mode-set!` — agent mode (interactive/plan/autopilot)
- `plan-read`, `plan-update!`, `plan-delete!` — session plan file ops
- `workspace-list-files`, `workspace-read-file`, `workspace-create-file!`
- `agent-list`, `agent-get-current`, `agent-select!`, `agent-deselect!`, `agent-reload!`
- `fleet-start!` — parallel sub-sessions

**Server-level** (`client.clj`):
- `mcp-config-list`, `mcp-config-add!`, `mcp-config-update!`, `mcp-config-remove!`

### Mock server enhancements

- `send-rpc-request!` — enables testing server→client RPCs (hooks, user-input, transforms)
- Response routing in server loop for bidirectional RPC testing
- 30+ new method stubs for all session/server RPC methods

### Instrumentation

Full `s/fdef` coverage and `instrument-all!`/`unstrument-all!` entries for all 19 new public functions.

### Before/After

| Metric | Before | After |
|---|---|---|
| Test cases | 118 | 145 (+27) |
| Assertions | 402 | 449 (+47) |
| Public API functions | ~85 | ~104 (+19) |

### Not included

Multi-client broadcast tests are deferred — they require significant mock server rework to support a two-client-one-session topology. These could be added as E2E tests in a follow-up.

### Commits

1. **Enhance mock server with server→client RPC support** — infrastructure
2. **Add test coverage for hooks, user-input, transforms, and tool results** — gap tests
3. **Add session RPC wrappers and integration tests** — new functions + tests + changelog

### CODE REVIEW

File: src/github/copilot_sdk/specs.clj [hunk 1]
```diff
@@ -625,9 +625,14 @@
           :opt-un [::selected-model ::reasoning-effort ::already-in-use? ::remote-steerable?
                    ::host-type ::head-commit ::base-commit]))
 
+(s/def ::status-code int?)
```
[the remark this answers is not in the document]
Comment (krukow) [reply src/github/copilot_sdk/specs.clj]: Fixed in 4de005c — changed `int?` to `integer?` which accepts both `Integer` and `Long`.

File: src/github/copilot_sdk/session.clj [hunk 2]
```diff
@@ -1130,6 +1130,141 @@
...
+  (let [{:keys [session-id client]} session
+        conn (connection-io client)]
+    (util/wire->clj
+     (proto/send-request! conn "session.mode.set" {:sessionId session-id :mode mode}))))
+
+;; -- Plan -------------------------------------------------------------------
+
+(defn ^:experimental plan-read
+  "Read the plan file for the session.
+   Returns a map with :exists? (boolean), :content (string or nil),
```
[the remark this answers is not in the document]
Comment (krukow) [reply src/github/copilot_sdk/session.clj:1157]: Fixed in 4de005c — `plan-read` now post-processes the result to rename `:exists` → `:exists?`, matching the documented API contract.

File: test/github/copilot_sdk/mock_server.clj [hunk 3]
```diff
@@ -463,3 +509,30 @@
   "Set context on a mock session (for testing list-sessions with context)."
   [server session-id context]
   (swap! (:sessions server) assoc-in [session-id :context] context))
+
+(defn send-rpc-request!
+  "Send a JSON-RPC request to the client and wait for the response.
+   Simulates server→client RPCs like hooks.invoke, userInput.request,
+   systemMessage.transform. Blocks until the client responds or timeout.
+   Must NOT be called from the mock server's server-loop thread.
+   Returns the response map (:result or :error)."
```
[the remark this answers is not in the document]
Comment (krukow) [reply test/github/copilot_sdk/mock_server.clj]: Fixed in 4de005c — docstring updated to: "Returns the full JSON-RPC response map on success. Throws ex-info on timeout or when the client responds with :error."

File: src/github/copilot_sdk/session.clj [hunk 4]
```diff
@@ -1130,6 +1130,141 @@
...
+     (proto/send-request! conn "session.agent.list" {:sessionId session-id}))))
+
+(defn ^:experimental agent-get-current
+  "Get the currently active custom agent for the session.
+   Returns a map with :name (string or nil)."
+  [session]
+  (let [{:keys [session-id client]} session
+        conn (connection-io client)]
+    (util/wire->clj
+     (proto/send-request! conn "session.agent.getCurrent" {:sessionId session-id}))))
```
[the remark this answers is not in the document]
Comment (krukow) [reply src/github/copilot_sdk/session.clj:1235]: Already addressed in e242d1f (prior commit) — `test-agent-get-current` and `test-agent-reload` integration tests were added.

File: src/github/copilot_sdk/client.clj [hunk 5]
```diff
@@ -1210,6 +1210,46 @@
...
+     (proto/send-request! conn "mcp.config.list" {}))))
+
+(defn ^:experimental mcp-config-add!
+  "Add an MCP server configuration.
+   params is a map with server config (name, command, args, etc.)."
+  [client params]
+  (ensure-connected! client)
+  (let [conn (:connection-io @(:state client))]
+    (util/wire->clj
+     (proto/send-request! conn "mcp.config.add" (util/clj->wire params)))))
```
[the remark this answers is not in the document]
Comment (krukow) [reply src/github/copilot_sdk/client.clj]: Fixed in 4de005c — removed the redundant `util/clj->wire` from `mcp-config-add!`, `mcp-config-update!`, and `mcp-config-remove!`. `proto/send-request!` handles wire conversion internally.

File: src/github/copilot_sdk/client.clj [hunk 6]
```diff
@@ -1210,6 +1210,46 @@
...
+     (proto/send-request! conn "mcp.config.update" (util/clj->wire params)))))
+
+(defn ^:experimental mcp-config-remove!
+  "Remove an MCP server configuration.
+   params is a map identifying the server to remove."
+  [client params]
+  (ensure-connected! client)
+  (let [conn (:connection-io @(:state client))]
+    (util/wire->clj
+     (proto/send-request! conn "mcp.config.remove" (util/clj->wire params)))))
```
[the remark this answers is not in the document]
Comment (krukow) [reply src/github/copilot_sdk/client.clj]: Already addressed in e242d1f — `test-mcp-config-add`, `test-mcp-config-update`, and `test-mcp-config-remove` integration tests were added.

File: test/github/copilot_sdk/mock_server.clj [hunk 7]
```diff
@@ -332,16 +370,23 @@
...
+          (try
+            (let [response (handle-request server msg)]
+              (write-message (:writer server) response))
+            (catch Exception e
+              (let [error-data (ex-data e)]
+                (write-message (:writer server)
+                               {:jsonrpc "2.0"
+                                :id (:id msg)
+                                :error {:code (or (:code error-data) -32603)
+                                        :message (.getMessage e)}}))))
```
[the remark this answers is not in the document]
Comment (krukow) [reply test/github/copilot_sdk/mock_server.clj]: Fixed in 4de005c — mock server now distinguishes requests (has `:id`) from notifications (has `:method` but no `:id`). Notifications are processed silently without writing a response.

File: src/github/copilot_sdk/session.clj [hunk 8]
```diff
@@ -1130,6 +1130,141 @@
...
+
+(defn ^:experimental plan-read
+  "Read the plan file for the session.
+   Returns a map with :exists? (boolean), :content (string or nil),
+   and :file-path (string or nil)."
+  [session]
+  (let [{:keys [session-id client]} session
+        conn (connection-io client)]
+    (util/wire->clj
+     (proto/send-request! conn "session.plan.read" {:sessionId session-id}))))
```
[the remark this answers is not in the document]
Comment (krukow) [reply src/github/copilot_sdk/session.clj]: Fixed in 4de005c — `plan-read` now renames `:exists` → `:exists?` in the return value, matching the docstring contract.

File: src/github/copilot_sdk/session.clj [hunk 9]
```diff
@@ -1130,6 +1130,141 @@
...
+
+(defn ^:experimental fleet-start!
+  "Start a fleet of parallel sub-sessions.
+   params is a map forwarded to the session.fleet.start RPC."
+  [session params]
+  (let [{:keys [session-id client]} session
+        conn (connection-io client)]
+    (util/wire->clj
+     (proto/send-request! conn "session.fleet.start"
+                          (merge {:sessionId session-id} (util/clj->wire params))))))
```
[the remark this answers is not in the document]
Comment (krukow) [reply src/github/copilot_sdk/session.clj]: Fixed in 4de005c — removed the redundant `util/clj->wire` from `fleet-start!`. Now passes raw params and lets `proto/send-request!` handle wire conversion.

File: doc/reference/API.md [hunk 10]
```diff
@@ -930,6 +952,32 @@ Get the client that owns this session.
...
 (session/mcp-disable! my-session "my-server")
+
+;; Get/set agent mode
+(session/mode-get my-session)
+;; => {:mode "interactive"}
+(session/mode-set! my-session "plan")
+
+;; Read/update session plan
+(session/plan-read my-session)
+;; => {:exists? true :content "# Plan\n..." :file-path "/path/to/plan.md"}
```
[the remark this answers is not in the document]
Comment (krukow) [reply doc/reference/API.md:963]: The implementation was fixed in 4de005c to rename `:exists` → `:exists?`, so the docs are now correct as-is.

File: test/github/copilot_sdk/integration_test.clj [hunk 11]
```diff
@@ -1625,3 +1625,532 @@
...
+  (testing "plan-read calls session.plan.read RPC"
+    (let [requests (atom [])
+          _ (mock/set-request-hook! *mock-server*
+              (fn [method params]
+                (swap! requests conj {:method method :params params})))
+          session (sdk/create-session *test-client*
+                    {:on-permission-request sdk/approve-all})]
+      (let [result (session/plan-read session)]
+        (is (some? result))
+        (is (some #(= "session.plan.read" (:method %)) @requests))))))
```
[the remark this answers is not in the document]
Comment (krukow) [reply test/github/copilot_sdk/integration_test.clj]: Fixed in 4de005c — `test-plan-read` now asserts `:exists?` key is present, its value is `false`, and `:content` is `nil`.

File: doc/reference/API.md [hunk 12]
```diff
@@ -486,6 +486,28 @@ Each quota snapshot map contains:
...
+;; => {:servers [...]}
+
+;; Add a new MCP server config
+(copilot/mcp-config-add! client {:name "my-server"
+                                  :command "npx"
+                                  :args ["-y" "@modelcontextprotocol/server-filesystem" "/tmp"]
+                                  :tools ["*"]})
+
+;; Update an existing config
+(copilot/mcp-config-update! client {:name "my-server" :tools ["read_file"]})
```
[the remark this answers is not in the document]
Comment (krukow) [reply doc/reference/API.md:505]: The `:mcp-command`/`:mcp-args`/`:mcp-tools` convention in session config is a Clojure-specific prefixing to avoid key collisions with other config fields in the same map. The `mcp-config-*` functions operate on standalone server config maps where there is no collision risk, so plain `:command`/`:args`/`:tools` keys are appropriate — they also match the upstream wire format more closely. I will add a brief note in the docstrings to clarify this distinction.

File: CHANGELOG.md [hunk 13]
```diff
@@ -3,6 +3,26 @@ All notable changes to this project will be documented in this file. This change
 
 ## [Unreleased]
 
+### Added
+- **Session RPC wrappers** — new experimental functions for session-level RPCs previously only accessible via `proto/send-request!`:
+  - `mode-get`, `mode-set!` — get/set agent mode (interactive/plan/autopilot)
+  - `plan-read`, `plan-update!`, `plan-delete!` — read/update/delete session plan file
+  - `workspace-list-files`, `workspace-read-file`, `workspace-create-file!` — session workspace file operations
+  - `agent-list`, `agent-get-current`, `agent-select!`, `agent-deselect!`, `agent-reload!` — custom agent CRUD
```
[the remark this answers is not in the document]
Comment (krukow) [reply CHANGELOG.md]: Good catch — fixed to say "custom agent management" instead of "CRUD".

File: .github/skills/update-upstream/SKILL.md [hunk 14]
```diff
@@ -0,0 +1,151 @@
...
+3. Run `bb validate-docs`
+
+### Phase 8: PR Creation
+
+1. Create a feature branch: `git checkout -b upstream-sync/v<version>`
+2. Commit changes in logical commits to make them easy to review commit by commit and with descriptive message and `Co-authored-by: Copilot <223556219+Copilot@users.noreply.github.com>`
+3. Push and create PR with `gh pr create`
+4. PR body should include: summary, changes list, validation results, review findings table
+
+### Phase 9: Reflecing on code review feedback.
```
[the remark this answers is not in the document]
Comment (krukow) [reply .github/skills/update-upstream/SKILL.md:111]: This typo is in `.github/skills/update-upstream/SKILL.md` which is not part of this PRs changes — it was picked up because the file is untracked. Will fix separately.

File: src/github/copilot_sdk/session.clj [hunk 15]
```diff
@@ -1130,6 +1130,146 @@
...
+
+(defn ^:experimental fleet-start!
+  "Start a fleet of parallel sub-sessions.
+   params is a map forwarded to the session.fleet.start RPC."
+  [session params]
+  (let [{:keys [session-id client]} session
+        conn (connection-io client)]
+    (util/wire->clj
+     (proto/send-request! conn "session.fleet.start"
+                          (merge {:session-id session-id} params)))))
```
[the remark this answers is not in the document]
Comment (krukow) [reply src/github/copilot_sdk/session.clj]: Fixed in c414af2 — `fleet-start!` now uses `assoc` after merge to force the session-id, preventing callers from overriding it via params.

File: test/github/copilot_sdk/mock_server.clj [hunk 16]
```diff
@@ -315,6 +316,43 @@
...
+                 "session.mcp.reload" {:success true}
+                 "session.extensions.list" {:extensions []}
+                 "session.extensions.enable" {:success true}
+                 "session.extensions.disable" {:success true}
+                 "session.extensions.reload" {:success true}
+                 "session.plugins.list" {:plugins []}
+                 "session.compaction.compact" {:success true}
+                 "session.shell.exec" {:exitCode 0 :stdout "" :stderr ""}
+                 "session.shell.kill" {:success true}
+                 "session.ui.elicitation" {:result {}}
```
[the remark this answers is not in the document]
Comment (krukow) [reply test/github/copilot_sdk/mock_server.clj]: Fixed in c414af2 — mock stub now returns `{:action "accept" :content {}}` matching the elicitation result shape.

File: CHANGELOG.md [hunk 17]
```diff
@@ -3,6 +3,26 @@ All notable changes to this project will be documented in this file. This change
...
+  - `workspace-list-files`, `workspace-read-file`, `workspace-create-file!` — session workspace file operations
+  - `agent-list`, `agent-get-current`, `agent-select!`, `agent-deselect!`, `agent-reload!` — custom agent management
+  - `fleet-start!` — start parallel sub-sessions
+- **MCP config wrappers** — new experimental server-level functions in `client`:
+  - `mcp-config-list`, `mcp-config-add!`, `mcp-config-update!`, `mcp-config-remove!` — MCP server configuration management
+- **Hooks integration tests** — 6 tests covering all hook types (preToolUse, postToolUse, sessionStart, unknownType, handler exceptions, no-hooks)
+- **User input handler tests** — 2 tests for `userInput.request` server→client RPC
+- **System message transform tests** — 3 tests for `systemMessage.transform` callback invocation, error fallback, and passthrough
+- **Tool result normalization tests** — 3 tests for string, nil, and structured ToolResultObject results via v3 broadcast
+- **Session RPC wrapper tests** — 13 integration tests for all new RPC wrapper functions
```
[the remark this answers is not in the document]
Comment (krukow) [reply CHANGELOG.md]: Fixed in c414af2 — updated count from 13 to 18 to reflect all wrapper tests added.

File: test/github/copilot_sdk/integration_test.clj [hunk 18]
```diff
@@ -1625,3 +1625,537 @@
...
+  (testing "fleet-start! calls session.fleet.start RPC"
+    (let [requests (atom [])
+          _ (mock/set-request-hook! *mock-server*
+              (fn [method params]
+                (swap! requests conj {:method method :params params})))
+          session (sdk/create-session *test-client*
+                    {:on-permission-request sdk/approve-all})]
+      (session/fleet-start! session {:prompt "do stuff"})
+      (let [rpcs (filter #(= "session.fleet.start" (:method %)) @requests)]
+        (is (= 1 (count rpcs)))))))
```
[the remark this answers is not in the document]
Comment (krukow) [reply test/github/copilot_sdk/integration_test.clj]: Fixed in af3af5c — `test-fleet-start` now passes a params map containing `:session-id "evil-override"` and asserts the actual wire `sessionId` is the real session ID (not the override). Also asserts `:prompt` is forwarded correctly.

File: src/github/copilot_sdk/client.clj [hunk 19]
```diff
@@ -1210,6 +1210,46 @@
...
+  [client]
+  (ensure-connected! client)
+  (let [conn (:connection-io @(:state client))]
+    (util/wire->clj
+     (proto/send-request! conn "mcp.config.list" {}))))
+
+(defn ^:experimental mcp-config-add!
+  "Add an MCP server configuration.
+   params is a map with server config (name, command, args, etc.)."
+  [client params]
```
[the remark this answers is not in the document]
Comment (krukow) [reply src/github/copilot_sdk/client.clj:1230]: Fixed in 5d6a531 — added explicit note to `mcp-config-add!`/`update!`/`remove!` docstrings clarifying they expect plain keys (`:name`, `:command`, `:args`, `:tools`) not the `:mcp-`-prefixed keys used in session config `:mcp-servers`.
